src/System.IO.UnmanagedMemoryStream/src/System/IO/UnmanagedMemoryAccessor.cs

```diff
@@ -418,11 +421,29 @@ public Decimal ReadDecimal(Int64 position)
                     int hi = UnsafeReadInt32(pointer + 8);
                     int flags = UnsafeReadInt32(pointer + 12);
 
-                    int* ptr = (int*)(&d);
-                    *ptr++ = flags;
-                    *ptr++ = hi;
-                    *ptr++ = lo;
-                    *ptr = mid;
+                    // Check for invalid Decimal values
+                    if (!((flags & ~(SignMask | ScaleMask)) == 0 && (flags & ScaleMask) <= (28 << 16)))
+                    {
+                        throw new IOException(SR.Arg_DecBitCtor);
+                    }
+
+                    int* ptr = (int*)&d;
+                    if (BitConverter.IsLittleEndian)
+                    {
+                        // Follow same order of fields declared in Decimal.cs
+                        *ptr++ = flags;
+                        *ptr++ = hi;
+                        *ptr++ = lo;
+                        *ptr = mid;
+                    }
+                    else
+                    {
+                        // Use reverse order
```

Comment:
And we don't have to swap the int bytes either, because everything in UnmanagedMemoryAccessor reads/writes in native endianness.